log4j2/src/test/java/com/google/common/flogger/backend/log4j2/Log4j2ScopedLoggingTest.java

```diff
@@ -0,0 +1,329 @@
+                .install()
+        ) {
+            GoogleLogContext logContext = (GoogleLogContext) googleLogger.atInfo();
+            logContext.log("test");
+            backend.log(logContext); // this event will be caught
+            Map<String, Object> contextMap = new HashMap<String, Object>();
+            contextMap.put("tags", "[foo, bar=baz2, bar=baz, anotherTag]");
+            assertLogCount(1);
+            assertLogEntry(0, INFO, "test", contextMap);
+        }
+    }
+
+    @Test
+    public void testTagsWithRepeatableMetadataClash() {
+        try (ScopedLoggingContext.LoggingContextCloseable ctx = ContextDataProvider.getInstance()
+                .getContextApiSingleton()
+                .newContext()
+                .withMetadata(REP_TAGS, "aValue")
+                .withMetadata(REP_TAGS, "anotherValue")
+                .withTags(Tags.builder().addTag("foo").addTag("bar", "baz").addTag("bar", "baz2").build())
+                .install()
+        ) {
+            GoogleLogContext logContext = (GoogleLogContext) googleLogger.atInfo();
+            logContext.log("test");
+            backend.log(logContext); // this event will be caught
+            Map<String, Object> contextMap = new HashMap<String, Object>();
+            contextMap.put("tags", "[foo, bar=baz2, bar=baz, [aValue, anotherValue]]");
+            assertLogCount(1);
+            assertLogEntry(0, INFO, "test", contextMap);
+        }
+    }
+
+    @Test
+    public void testMultipleRepeatableMetadata() {
+        try (ScopedLoggingContext.LoggingContextCloseable ctx = ContextDataProvider.getInstance()
+                .getContextApiSingleton()
+                .newContext()
+                .withMetadata(REP_KEY, 1)
+                .withMetadata(REP_KEY, 2)
+                .install()
+        ) {
+            GoogleLogContext logContext = (GoogleLogContext) googleLogger.atInfo();
+            logContext.log("test");
+            backend.log(logContext); // this event will be caught
+            Map<String, Object> contextMap = new HashMap<String, Object>();
+            contextMap.put("rep", "[1, 2]");
+            assertLogCount(1);
+            assertLogEntry(0, INFO, "test", contextMap);
+        }
+    }
+
+    @Test
+    public void testMultipleNonRepeatableMetadataSameMetadataKey() {
+        try (ScopedLoggingContext.LoggingContextCloseable ctx = ContextDataProvider.getInstance()
+                .getContextApiSingleton()
+                .newContext()
+                .withMetadata(ID_KEY, "001")
+                .withMetadata(ID_KEY2, "002")
+                .install()
+        ) {
+            GoogleLogContext logContext = (GoogleLogContext) googleLogger.atInfo();
+            logContext.log("test");
+            backend.log(logContext); // this event will be caught
+            Map<String, Object> contextMap = new HashMap<String, Object>();
+            contextMap.put("id", "[002, 001]");
```

Comment:
Hmm. Not what I'd have expected.

Metadata preserves order, and so does MetadataProcessor, so I'd have expected "[001, 002]".

*doh* This is due to the linked list acting like a stack and reversing things. That's my fault for not spotting.
Damn, I think this means the list needs to be redone to append new items to the end, not prepend.

This means you need a "list" type and a separate "node" type, at which point I'd just use `java.util.LinkedList` to be honest (unless you need to see a unique type later in the log4j2 handlers).


Sorry, my bad.

Reply:
> Sorry, my bad.

It's alright :-) We are trying to find the most suitable solution and I certainly did my fair share of making things a bit more cumbersome then they need to be. I hope my recent changes help to get a step closer to meeting expectations.

> In theory someone could make a custom metadata key using the Tags type

I checked this use case and realized that there was another bug in my logic of handling metadata. 
Consider the following Tags definition:
```
Given:
Tags.builder().addTag("foo", "bar").build()

this formats to:
tags=foo=bar
```
I did not expect it. To fix this problem (and the version of using a custom Tag) as well as using Tags as MetadataKeys, a bit more work was required.

In order to fix the order of stored metadata, I replaced the value list with a simple FIFO linked-list implementation and basically kept the `toString()` method as defined previously. I might be wrong, but I find it a bit more intuitive if the `toString()` method returns something like "" (empty string) - empty queue; a - queue contains value a; [a, b] - queue contains two values a and b instead of [], a and [a, b]. The use case of an empty queue should not happen and is only considered to avoid an NPE.

Achieving consistent formatting especially when mixing metadata (single and repeatable), tags with single and multiple values required a few changes to the code you already reviewed (Log4j2SimpleLogEvent). I think I got it resolved by distinguishing tags (LogContext.Key.TAGS and Tags) from "standard" metadata. Tags are always formatted as a list whereas "standard" metadata is usually formatted as a key, a key/value-pair or a list. I hope the test cases cover most scenarios.

Finally, I removed the custom message formatter and instead copied a slightly modified version of Flogger's SimpleMessageFormatter that ignores metadata when formatting the message.

Reply:
I replaced the custom linked list implementation used in the ValueQueue with `LinkedList`. As you and @cgdecker pointed out using a standard type is sufficient.